### PR TITLE
Add `-fno-delete-null-pointer-checks` and friends

### PR DESCRIPTION
**Pull request by david-a-wheeler**
Add the following:

* `-fno-delete-null-pointer-checks`
* `-fno-strict-overflow`
* `-fno-strict-aliasing`
* `-ftrivial-auto-var-init`

This is a redo/re-implementation of this pull request, per various comments:
https://github.com/ossf/wg-best-practices-os-developers/pull/139

### CODE REVIEW

**Comment by david-a-wheeler** [comment]
Note: There are a number of unrelated markdownlint errors.  We should fix them, but I recommend fixing them in a *separate* branch since they are unrelated to this proposal.

**Comment by david-a-wheeler** [comment]
@[thomasnyman](https://github.com/thomasnyman) @[gkunz](https://github.com/gkunz) - here are several new flags that start using the "production code vs. instrumented test code" distinction. Now that we have a path forward, let's add some flags that use it.

**Comment by david-a-wheeler** [comment]
I don't know when these flags were added in GCC and Clang. I know that others are really good at finding out!

**Comment by david-a-wheeler** [comment]
This also fixes: https://github.com/ossf/wg-best-practices-os-developers/issues/267

I think the problem earlier was we were trying to use the same options for *both* production code *and* for instrumented test code. Once you split things up, I think it's clearer.

File: docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md [hunk 1]
```diff
@@ -37,6 +37,8 @@ When compiling code in any of the situations in the below table, add the corresp
 | for shared libraries | `-fPIC -shared`                |
 | for x86_64           | `-fcf-protection=full`         |
 | for aarch64          | `-mbranch-protection=standard` |
+| for instrumented test code | `-ftrivial-auto-var-init=pattern` |
```
Comment (thomasnyman) [inline docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:40]: As`-ftrivial-auto-var-init` interferes with unitialized memory tracking in tools such as Valgrind and AddressSanitizer I don't. think it should be enabled for instrumented test code. There was a discussion on this in [this comment to #245](https://github.com/ossf/wg-best-practices-os-developers/issues/245#issuecomment-1785479519).

Comment (thomasnyman) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:40]: @thesamesam Thanks, I was thinking of MSAN too, not ASAN. The `-ftrivial-auto-var-init` does interfere with MSAN as can be seen here: https://godbolt.org/z/6qTPz9n6h

Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:40]: If you don't use MSAN, is this still best to not use? I think many developers don't use MSAN.

Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:40]: It'd be easy to remove the `=pattern` case. That's probably the best strategy, at least for now.

Comment (david-a-wheeler) [reply docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md:40]: Ok, let's do that. We can discuss =pattern in the details, without recommending its use in particular.

**Comment by david-a-wheeler** [comment]
I believe everything is ready to go in this pull request.

I propose merging this. Is there agreement?

**Pull request merged**